operator/v1alpha1/types_kueue.go

```diff
@@ -0,0 +1,324 @@
+)
+
+// +kubebuilder:validation:Enum="Enabled;Disabled"
+type EnabledOrDisabled string
+
+const (
+	Enabled  EnabledOrDisabled = "Enabled"
+	Disabled EnabledOrDisabled = "Disabled"
+)
+
+type KueueConfiguration struct {
+	// waitForPodsReady configures gang admission
+	// +optional
+	WaitForPodsReady WaitForPodsReady `json:"waitForPodsReady,omitempty"`
+	// integrations are the types of integrations Kueue will manager
+	// +required
+	Integrations Integrations `json:"integrations"`
+	// featureGates are advanced features for Kueue
+	// +optional
+	FeatureGates map[string]EnabledOrDisabled `json:"featureGates,omitempty"`
+	// resources provides additional configuration options for handling the resources.
+	// Supports https://github.com/kubernetes-sigs/kueue/blob/release-0.10/keps/2937-resource-transformer/README.md
+	// +optional
+	Resources Resources `json:"resources,omitempty"`
+	// manageJobsWithoutQueueName controls whether or not Kueue reconciles
+	// jobs that don't set the annotation kueue.x-k8s.io/queue-name.
+	// Allowed values are NoQueueName and QueueName
+	// Default will be QueueName
+	// +kubebuilder:default=QueueName
+	// +optional
+	ManageJobsWithoutQueueName *ManageJobsWithoutQueueNameOption `json:"manageJobsWithoutQueueName,omitempty"`
+	// managedJobsNamespaceSelector can be used to omit some namespaces from ManagedJobsWithoutQueueName
+	// Only valid if ManagedJobsWithoutQueueName is NoQueueName
+	// +optional
+	ManagedJobsNamespaceSelector *metav1.LabelSelector `json:"managedJobsNamespaceSelector,omitempty"`
+	// fairSharing controls the fair sharing semantics across the cluster.
+	// +optional
+	FairSharing FairSharing `json:"fairSharing,omitempty"`
+	// metrics allows one to change if metrics
+	// are enabled or disabled.
+	// Microshift does not enable metrics by default
+	// Default will assume metrics are enabled.
+	// +kubebuilder:default=Enabled
+	// +optional
+	Metrics *EnabledOrDisabled `json:"metrics,omitempty"`
+}
+
+// KueueStatus defines the observed state of Kueue
+type KueueStatus struct {
+	operatorv1.OperatorStatus `json:",inline"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// KueueList contains a list of Kueue
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type KueueList struct {
+	metav1.TypeMeta `json:",inline"`
+	// metadata for the list
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty"`
+	// items is a slice of kueue
+	// this is a cluster scoped resource and there can only be 1 kueue
+	// +kubebuilder:validation:MaxItems=1
+	// +required
+	Items []Kueue `json:"items"`
+}
+
+// These structs come directly from Kueue.
+type Resources struct {
+	// excludeResourcePrefixes defines which resources should be ignored by Kueue
```

Comment:
Your godoc wants to explain what a valid resource prefix is, I don't even know what an example of this might be. As someone new to this API, I wouldn't know what I'm supposed to do with this field.

Reply:
It's pretty unconstrained...any string is legal (although many will have no effect).  The only place excludedResourcePrefixes  is used in Kueue (if you ignore the plumbing to get it from the config file to the place it is used) is as an argument to strings.hasPrefix in [dropExcludedResources](https://github.com/kubernetes-sigs/kueue/blob/main/pkg/workload/workload.go#L267-L282).

Reply:
But what are the `resources` that are described here? Thinking about this from the end user perspective, I'm assuming by resource you mean some sort of API resource? For example a `deployments.apps` resource?

Reply:
So then this is matching against `corev1.ResourceName` IIUC, which has the following restriction:
> Resource names must be not more than 63 characters, consisting of upper- or lower-case alphanumeric characters, with the -, _, and . characters allowed anywhere, except the first or last character.

Reply:
Is there an objection to not including that validation here for now?

This is more of an advanced feature for folks using Instaslice integration atm.

Kueue does not have strict validation on this and this is not yet a GA feature.

Reply:
Is there a strong reason for _not_ including the validation? 

We generally encourage the validations to ensure users get early feedback on values we _know_ are going to be invalid. It results in a better user experience overall to reject at admission time using validations on the CRD. If Kueue is using Kubernetes libraries for this logic, I think enforcing what Kubernetes does is reasonable.

"Because X doesn't do this" usually isn't a strong enough argument IMO. If there is an anticipated feature that will contradict this validation that is something that is a reasonable argument.

Reply:
So I agree with you but I am not sure Kueue is validating these items as strict as you want.

So I tend to push back on making this more strict.

If we want to drive change in this API I think it should come from upstream and down. And not the other way around.

This is a beta feature at the moment and the general idea of these "transformations" was very vague. Kueue maintainers had some ideas about tiing these validation transformations to cost or other arbitrary metric.

Right now, Kueue requires resourceLists but I think their only validation may be in go marshaling AFAIK.

Reply:
https://github.com/kubernetes-sigs/kueue/blob/main/pkg/config/validation.go#L304

best I can do is enforce uniqueness?